File: src/csharp/3_StructuralPatterns/1_Adapter/NoCaching.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq.Expressions;
using System.Security.AccessControl;
using MoreLinq;
using static System.Console;

namespace DotNetDesignPatternDemos.Structural.Adapter.NoCaching
{
  public class Point
  {
    public int X;
    public int Y;

    public Point(int x, int y)
    {
      this.X = x;
      this.Y = y;
    }

    public override string ToString()
    {
      return $"{nameof(X)}: {X}, {nameof(Y)}: {Y}";
    }
  }

  public class Line
  {
    public Point Start;
    public Point End;

    public Line(Point start, Point end)
    {
      this.Start = start;
      this.End = end;
    }
  }

  public class VectorObject : Collection<Line>
  {
    
  }

  public class VectorRectangle : VectorObject
  {
    public VectorRectangle(int x, int y, int width, int height)
    {
      Add(new Line(new Point(x,y), new Point(x+width, y) ));
      Add(new Line(new Point(x+width,y), new Point(x+width, y+height) ));
      Add(new Line(new Point(x,y), new Point(x, y+height) ));
      Add(new Line(new Point(x,y+height), new Point(x+width, y+height) ));
    }
  }

  public class LineToPointAdapter : Collection<Point>
  {
    private static int count = 0;

    public LineToPointAdapter(Line line)
    {
      WriteLine($"{++count}: Generating points for line [{line.Start.X},{line.Start.Y}]-[{line.End.X},{line.End.Y}] (no caching)");

      int left = Math.Min(line.Start.X, line.End.X);
      int right = Math.Max(line.Start.X, line.End.X);
      int top = Math.Min(line.Start.Y, line.End.Y);
      int bottom = Math.Max(line.Start.Y, line.End.Y);
      int dx = right - left;
      int dy = line.End.Y - line.Start.Y;

      if (dx == 0)
      {
        for (int y = top; y <= bottom; ++y)
        {
          Add(new Point(left, y));
        }
      } else if (dy == 0)
      {
        for (int x = left; x <= right; ++x)
        {
          Add(new Point(x, top));
        }
      }
    }
  }

  public class Demo
  {
    private static readonly List<VectorObject> vectorObjects = new List<VectorObject>
    {
      new VectorRectangle(1, 1, 10, 10),
      new VectorRectangle(3, 3, 6, 6)
    };

    // the interface we have
    public static void DrawPoint(Point p)
    {
      Write(".");
    }

    static void Main()
    {
      Draw();
      Draw();
    }

    private static void Draw()
    {
      foreach (var vo in vectorObjects)
      {
        foreach (var line in vo)
        {
          var adapter = new LineToPointAdapter(line);
          adapter.ForEach(DrawPoint);
        }
      }
    }
  }
}

File: src/java/3_StructuralPatterns/1_Adapter/NoCaching.java
package com.activemesa.structural.adapter.nocaching;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Point
{
  public int x, y;

  public Point(int x, int y)
  {
    this.x = x;
    this.y = y;
  }

  @Override
  public String toString()
  {
    return "Point{" +
      "first=" + x +
      ", second=" + y +
      '}';
  }
}

class Line
{
  public Point start, end;

  public Line(Point start, Point end)
  {
    this.start = start;
    this.end = end;
  }
}

class VectorObject extends ArrayList<Line> {}

class VectorRectangle extends VectorObject
{
  public VectorRectangle(int x, int y, int width, int height)
  {
    add(new Line(new Point(x,y), new Point(x+width, y) ));
    add(new Line(new Point(x+width,y), new Point(x+width, y+height) ));
    add(new Line(new Point(x,y), new Point(x, y+height) ));
    add(new Line(new Point(x,y+height), new Point(x+width, y+height) ));
  }
}

class LineToPointAdapter extends ArrayList<Point>
{
  private static int count = 0;

  public LineToPointAdapter(Line line)
  {
    System.out.println(
      String.format("%d: Generating points for line [%d,%d]-[%d,%d] (no caching)",
        ++count, line.start.x, line.start.y, line.end.x, line.end.y));

    int left = Math.min(line.start.x, line.end.x);
    int right = Math.max(line.start.x, line.end.x);
    int top = Math.min(line.start.y, line.end.y);
    int bottom = Math.max(line.start.y, line.end.y);
    int dx = right - left;
    int dy = line.end.y - line.start.y;

    if (dx == 0)
    {
      for (int y = top; y <= bottom; ++y)
      {
        add(new Point(left, y));
      }
    }
    else if (dy == 0)
    {
      for (int x = left; x <= right; ++x)
      {
        add(new Point(x, top));
      }
    }
  }
}

class AdapterDemo
{
  private static final List<VectorObject> vectorObjects =
    new ArrayList<>(Arrays.asList(
      new VectorRectangle(1,1,10,10),
      new VectorRectangle(3,3,6,6)
    ));

  public static void drawPoint(Point p)
  {
    System.out.print(".");
  }

  private static void draw()
  {
    for (VectorObject vo : vectorObjects)
    {
      for (Line line : vo)
      {
        LineToPointAdapter adapter = new LineToPointAdapter(line);
        adapter.forEach(AdapterDemo::drawPoint);
      }
    }
  }

  public static void main(String[] args)
  {
    draw();
    draw();
  }
}
